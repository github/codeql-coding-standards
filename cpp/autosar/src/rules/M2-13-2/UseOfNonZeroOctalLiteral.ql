/**
 * @id cpp/autosar/use-of-non-zero-octal-literal
 * @name M2-13-2: Octal constants (other than zero) shall not be used
 * @description Use of octal constants (other than zero) can lead to confusion with decimal numbers.
 * @kind problem
 * @precision very-high
 * @problem.severity recommendation
 * @tags external/autosar/id/m2-13-2
 *       readability
 *       external/autosar/allocated-target/architecture
 *       external/autosar/allocated-target/design
 *       external/autosar/allocated-target/implementation
 *       external/autosar/enforcement/automated
 *       external/autosar/obligation/required
 */

import cpp
import codingstandards.cpp.autosar
import codingstandards.cpp.Cpp14Literal

from Cpp14Literal::OctalLiteral octalLiteral
where
  not isExcluded(octalLiteral, LiteralsPackage::useOfNonZeroOctalLiteralQuery()) and
  not octalLiteral.getValue() = "0"
select octalLiteral, "Non zero octal literal " + octalLiteral.getValueText() + "."
