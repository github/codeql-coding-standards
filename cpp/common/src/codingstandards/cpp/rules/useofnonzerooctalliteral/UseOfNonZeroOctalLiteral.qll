/**
 * Provides a library with a `problems` predicate for the following issue:
 * Octal constants shall not be used.
 */

import cpp
import codingstandards.cpp.Customizations
import codingstandards.cpp.Exclusions
import codingstandards.cpp.Cpp14Literal

abstract class UseOfNonZeroOctalLiteralSharedQuery extends Query { }

Query getQuery() { result instanceof UseOfNonZeroOctalLiteralSharedQuery }

query predicate problems(Cpp14Literal::OctalLiteral octalLiteral, string message) {
  not isExcluded(octalLiteral, getQuery()) and
  not octalLiteral.getValue() = "0" and
  message = "Non zero octal literal " + octalLiteral.getValueText() + "."
}
