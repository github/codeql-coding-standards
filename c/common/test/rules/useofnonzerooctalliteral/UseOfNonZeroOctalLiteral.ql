// GENERATED FILE - DO NOT MODIFY
import codingstandards.cpp.rules.useofnonzerooctalliteral.UseOfNonZeroOctalLiteral

class TestFileQuery extends UseOfNonZeroOctalLiteralSharedQuery, TestQuery { }
