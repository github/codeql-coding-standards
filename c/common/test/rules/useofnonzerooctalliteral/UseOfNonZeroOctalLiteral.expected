| test.c:5:3:5:5 | 10 | Non zero octal literal 012. |
| test.c:6:3:6:5 | 44 | Non zero octal literal 054. |
